### PR TITLE
Implement message signing on Key

### PR DESCRIPTION
**Pull request by sisou**
This PR improves the `signMessage()` function on the Key class. It does not yet implement the request type.

### CODE REVIEW

File: Nimiq.d.ts [hunk 1]
```diff
@@ -41,7 +41,35 @@ declare namespace Nimiq {
...
 
-    class SerialBuffer extends Uint8Array {}
+    class SerialBuffer extends Uint8Array {
+        constructor(bufferOrArrayOrLength: any)
+        subarray(start: number, end: number): Uint8Array
+        readPos: number
+        writePos: number
+        reset(): void
+        read(length: number): Uint8Array
+        write(array: any): void
```
Comment (Bettelstab) [inline Nimiq.d.ts]: any => SerialBuffer?

File: Nimiq.d.ts [hunk 2]
```diff
@@ -41,7 +41,35 @@ declare namespace Nimiq {
         static toHex(buf: Uint8Array): string
     }
 
-    class SerialBuffer extends Uint8Array {}
+    class SerialBuffer extends Uint8Array {
+        constructor(bufferOrArrayOrLength: any)
```
Comment (Bettelstab) [inline Nimiq.d.ts]: any =>  SerialBuffer | number ?

File: Nimiq.d.ts [hunk 3]
```diff
@@ -41,7 +41,35 @@ declare namespace Nimiq {
...
+        subarray(start: number, end: number): Uint8Array
+        readPos: number
+        writePos: number
+        reset(): void
+        read(length: number): Uint8Array
```
Comment (Bettelstab) [inline Nimiq.d.ts]: Unint8Array => SerialBuffer?

File: Nimiq.d.ts (hunk 1 above)
Comment (sisou) [reply Nimiq.d.ts]: I took the annotations directly from the core, and they are using `{*}`. So we have to honor that here.

File: Nimiq.d.ts (hunk 3 above)
Comment (sisou) [reply Nimiq.d.ts]: No, the core is returning a `Uint8Array` here.

File: Nimiq.d.ts (hunk 2 above)
Comment (sisou) [reply Nimiq.d.ts]: Well, it has the same options as a Uint8Array: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array

File: src/lib/Key.js [hunk 4]
```diff
@@ -78,14 +78,48 @@ class Key { // eslint-disable-line no-unused-vars
...
     signMessage(message) {
-        message = `nimiq_msg_${message}`;
-        const msgBytes = Utf8Tools.stringToUtf8ByteArray(message);
-        const signature = Nimiq.Signature.create(this.keyPair.privateKey, this.keyPair.publicKey, msgBytes);
-        return { message, signature };
+        /**
+         * Adding a prefix to the message makes the calculated signature recognisable as
+         * a Nimiq specific signature. This prevents misuse where a malicious request can
+         * sign arbitrary data (e.g. a transaction) and use the signature to impersonate
+         * the victim. (https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_sign)
```
Comment (danimoh) [inline src/lib/Key.js]: :+1:

File: src/lib/Key.js [hunk 5]
```diff
@@ -78,14 +78,48 @@ class Key { // eslint-disable-line no-unused-vars
...
+
+        const bufLength = 1 /* prefixBytes length value */
+            + prefixLength
+            + 1 /* msgBytes length value */
+            + msgLength;
+
+        // Construct buffer
+        const buf = new Nimiq.SerialBuffer(bufLength);
+        buf.writeUint8(prefixLength);
+        buf.write(prefixBytes);
```
Comment (danimoh) [inline src/lib/Key.js]: There is also SerialBuffer.writeVarLengthStringth

Comment (sisou) [reply src/lib/Key.js]: Yes, but it only supports ASCII characters and we want to support the full UTF-8 range.

Comment (sisou) [reply src/lib/Key.js]: Also, I need to know the length of the prefix string anyway to be able to size the buffer (yes, hardcoding would work but that makes it more inflexible).

**Pull request merged**